src/modules/roc_core/target_posix/roc_core/secure_random.cpp

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright (c) 2020 Roc authors
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#include <stdlib.h>
+#include <uv.h>
+
+#include "roc_core/fast_random.h"
+#include "roc_core/panic.h"
+#include "roc_core/secure_random.h"
+
+namespace roc {
+namespace core {
+
+// Based on uv_random from libuv and arc4random_uniform() from OpenBSD.
+uint32_t secure_random(uint32_t from, uint32_t to) {
+#if UV_VERSION_MAJOR >= 1 && UV_VERSION_MINOR >= 33
+    roc_panic_if_not(from <= to);
+
+    const uint64_t upper = (uint64_t)to - from + 1;
+    const uint64_t min = -upper % upper;
+    uint64_t val;
+
+    for (;;) {
+        if (!uv_random(NULL, NULL, &val, sizeof(val), 0, NULL) && val >= min) {
+            break;
+        }
+    }
```

Comment:
So if uv_random() will fail every time (return non-zero), e.g. because /dev/urandom is not available, this loop will hang.

Please always do proper error checking. If uv_random() returns an error, we should log the error message (you can find examples in roc_netio) and return false to the upper level.

The code that calls secure_random() should handle the error as well and fail the session construction. I already mentioned it in the issue, actually: https://github.com/roc-project/roc/issues/102#issuecomment-605116491

Reply:
cool, to achieve this, what I can think of is change the definition to `int secure_random(uint32_t from, uint32_t to, unsigned &to_assign);` and pass source_, seqnum_, timestamp_ to to_assign and return succeed or not, do you have a better way to indicate the status as well as generate the random number? thanks!

Reply:
or make it as uint32_t* and NULL indicates failure?

Reply:
```
uint32_t* ptr = new uint32_t;
*ptr = ret;
return ptr;
```

This is a very bad design decision on many levels:

1. We return a raw pointer and transfer ownership to the caller. It's easy to forget to deallocate it and get a memory leak. The function signature gives no hint on this (e.g. it could be named "alloc_secure_random"), and you even didn't mention such an important thing in the function comment.

2. The preferred way to *transfer ownership* is to return a smart pointer, which will save the caller from producing memory leaks. There are many cases when raw pointers are still useful, but they should be used with care.

    If you're using `delete` in your code and this code does not belong to some low-level memory management subsystem (e.g. you're writing an allocator), it's a hint that you're probably do something wrong. In most cases you just need a smart pointer.

    In this project, we almost never use raw `new` calls as well. Instead, we almost always allocate memory either from `core::Pool` or from `core::IAllocator` and combine them with smart pointers (`core::ScopedPtr` and `core::SharedPtr`).

3. But anyway, there is absolutely no need to allocate anything here at all. This is an unreasonable burden both to the programmer who uses your function and to performance (allocations are not cheap).

    There is nothing wrong with your first suggestion and it's a common practice in C:

        bool secure_random(uint32_t from, uint32_t to, uint32_t& result);

    (Note that in this project we mostly use `bool` instead of `int` to indicate a failure).

Reply:
thank you for your advice! always learning from you!